### PR TITLE
Meltdown: check kernel args

### PR DESCRIPTION
**Pull request by JackSlateur**
If pti=off or nopti is passed as a kernel argument, kpti is disabled, so even if
the kernel supports it, the system is vulnerable to variant 3
The cmdline check has the highest priority, but is disabled if
/sys/kernel/debug/x86/pti_enabled is found, as this is the only method to enable
kpti at runtime (as far as I know)

Signed-off-by: Alexandre Bruyelles <abruyelles@odiso.com>

### CODE REVIEW

**Comment by speed47** [comment]
I've just tested on a RedHat/CentOS kernel (I think that's what you're using given the `/sys/kernel/debug/x86/pti_enabled` path your referring to which is specific to them), and if I pass `noibrs noibpb nopti` on the kernel command line, all these features are disabled on boot indeed, but I can still re-enable these features at runtime by echoing the proper value to the proper file in the `/sys/kernel/debug/x86` directory.

In other words it's not necessary AFAICT to grep for these options on the command-line, as the current status of the mitigations is always available in the  `/sys/kernel/debug/x86/{ibrs,ibpb,pti}_enabled` files.

I'm closing this PR but please feel free to reopen and comment if you found a different behavior on your system. Thanks!

**Comment by JackSlateur** [comment]
Hi,

I am using Debian, kernel 4.13.13 (this does not have the /sys/devices/system/cpu/vulnerabilities/* interface)

I disabled the cmdline grep if /sys/kernel/debug/x86/pti_enabled is found because this file override the kernel argument, and I want not to create bugs for the redhat folks :)

So, when I set nopti as a kernel argument, the scripts say that I am protected (which I am not)

Thank

**Comment by speed47** [comment]
I downloaded the 4.13.13 kernel from debian, ran my script on it, and I think I know what happened in your case. Could you try the `dmesg_truncated` branch and report the results?
( https://raw.githubusercontent.com/speed47/spectre-meltdown-checker/dmesg_truncated/spectre-meltdown-checker.sh ).

**Comment by JackSlateur** [comment]
For a newly rebooted system, your branch works fine (`PTI enabled and active:  NO`)
For another system, it outputs an unknown status (`PTI enabled and active:  UNKNOWN  (dmesg truncated, please reboot and relaunch this script)`), which is `VULNERABLE` in the batch outputs

So, this is not perfect, but covers my needs, thank you :)

**Comment by speed47** [comment]
The dmesg test is used as a last resort if every other way of checking if PTI (or KAISER) is enabled, failed.
So, in the end, it most probably means that PTI/KAISER is not enabled on your system, which is why the status says "vulnerable". If I were to put "unknown" here, it would mean that systems would change from "vulnerable" to "unknown" when the dmesg ring buffer overwrite the first boot messages... And I don't want people's monitoring to turn red just because of that as the uptime is growing !

Merging the dmesg_truncated branch.
